src/mame/sinclair/spec128.cpp

```diff
@@ -358,26 +433,35 @@ void spectrum_128_state::spectrum_128(machine_config &config)
 	m_maincpu->set_m1_map(&spectrum_128_state::spectrum_128_fetch);
 	m_maincpu->set_vblank_int("screen", FUNC(spectrum_128_state::spec_interrupt));
 	m_maincpu->nomreq_cb().set(FUNC(spectrum_128_state::spectrum_nomreq));
+	m_maincpu->busack_cb().set(m_dma, FUNC(z80dma_device::bai_w));
 
 	config.set_maximum_quantum(attotime::from_hz(60));
 
-	/* video hardware */
+	Z80DMA(config, m_dma, X1_128_SINCLAIR / 10);
+	m_dma->out_busreq_callback().set_inputline(m_maincpu, Z80_INPUT_LINE_BUSRQ);
+	m_dma->out_int_callback().set_inputline(m_maincpu, INPUT_LINE_IRQ0);
+	m_dma->in_mreq_callback().set([this](offs_t offset) { return m_maincpu->space(AS_PROGRAM).read_byte(offset); });
+	m_dma->out_mreq_callback().set([this](offs_t offset, u8 data) { m_maincpu->space(AS_PROGRAM).write_byte(offset, data); });
+	m_dma->in_iorq_callback().set([this](offs_t offset) { return m_maincpu->space(AS_IO).read_byte(offset); });
+	m_dma->out_iorq_callback().set([this](offs_t offset, u8 data) { m_maincpu->space(AS_IO).write_byte(offset, data); });
```

Comment:
Should this really be here and not connected by default, or should the modified system with the DMA controller be a clone?

Also, you need an input merger when multiple things are trying to drive the same input (`INPUT_LINE_IRQ0` in this case).

Reply:
I don't think clone approach is feasible. In case we have multiple custom MODs: TurboFM, DMA, and potentially something else - clone approach is inachievable.
I better prefer to add these MODs as exp devices which they are not - problem with exp slot is... not all the clone have it.

Reply:
Would fake slots to represent modifications be feasible?  Just throwing ideas around.

Reply:
@cuavas I agree with all your concerns. Unfortunately I don't see straightforward way to do it now.
1. Input Merger required to revisit all spectrum's children which use inheritance.
2. Moving installation to startup: I don't know the way to access config params in these phases.

Putting this on my TODO list:
1. Apply input_merger for INPUT_LINE_IRQ0 in spectrum hierarchy
2. Investigate possibility of lightway slots for custom mods